app/controllers/application_controller.rb

```diff
@@ -100,4 +100,30 @@ def token
   def cache
     params[:cache] ? { cache: params[:cache] } : {}
   end
+
+  def set_no_cache_headers
+    response.headers["Cache-Control"] = "no-store"
+  end
+
+  def fetch_http_referrer
+    http_referrer = request.headers["HTTP_REFERER"]
+
+    return nil unless http_referrer&.start_with?(Plek.new.website_root)
+
+    http_referrer.delete_prefix Plek.new.website_root
+  end
+
+  def redirect_with_ga(url, ga_client_id = nil)
+    ga_client_id ||= params[:_ga]
+    if ga_client_id
+      url =
+        if url.include? "?"
+          "#{url}&_ga=#{ga_client_id}"
+        else
+          "#{url}?_ga=#{ga_client_id}"
+        end
```

Comment:
We already have Addressable as a dependency so we could use it here
```suggestion
      uri = Addressable::URI.parse(url)
      uri.query_values = (uri.query_values || {}).merge("_ga" => ga_client_id)
      url = uri.to_s
```

Reply:
We use this in [other apps](https://github.com/alphagov/collections-publisher/blob/main/app/presenters/coronavirus/sub_section_json_presenter.rb#L45-L53) too (thank you @leenagupte for showing me this) so I think extracting it would be good. I'm not sure `govuk_personalisation` is the right place though... 🤔

Reply:
Ace, ok if it's place is not in the gem then here sounds good.
I've put it in its own prior commit, that is separate from the move into `ApplicationController`

See here:
https://github.com/alphagov/frontend/pull/2781/commits/ac726d7509520a95378f7020912d502caff7d529

Thanks both @AlanGabbianelli and @leenagupte !